rmgpy/data/kinetics/family.py

```diff
@@ -2059,68 +2063,86 @@ def getLabeledReactantsAndProducts(self, reactants, products):
-        products = reaction.products[:]
-       
-        labeledReactants, labeledProducts = self.getLabeledReactantsAndProducts(reactants, products)
-        if not labeledReactants and not labeledProducts:
-            if len(reactants) > 1:
-                # Check for swapped reactants if there are more than 1
-                labeledReactants, labeledProducts = self.getLabeledReactantsAndProducts(list(reversed(reactants)), products)
-        if not labeledReactants and not labeledProducts:
-            raise Exception("RMG could not label atoms for this reaction in {}.".format(self.label))
-        
-        labeledProducts_spcs = []
-        labeledReactants_spcs = []
-        for labeledProduct in labeledProducts:
-            labeledProducts_spcs.append(Species(molecule=[labeledProduct]))
-        reaction.products = labeledProducts_spcs
-        for labeledReactant in labeledReactants:
-            labeledReactants_spcs.append(Species(molecule=[labeledReactant]))
-        reaction.reactants = labeledReactants_spcs
+            raise IndexError('You have {0} reactants, which is unexpected!'.format(len(reactants)))
+        
+    def addAtomLabelsForReaction(self, reaction, output_with_species = True):
+        """
+        Apply atom labels on a reaction using the appropriate atom labels from
+        this reaction family.
+
+        The reaction is modified in place, containing species objects if
+        output_with_species is True and containing molecule objects, which
+        are the exact resonance structures used for the reaction, if
+        output_with_species is False. None is returned.
+        """
+        # make sure we start with reaction with species objects
+        reaction.ensure_species(reactant_resonance=False, product_resonance=False)
+
+        # get all possible pairs of resonance structures
+        reactant_pairs = list(itertools.product(*[s.molecule for s in reaction.reactants]))
+        product_pairs = list(itertools.product(*[s.molecule for s in reaction.products]))
+
+        labeled_reactants, labeled_products = None, None
+        # go through each combination of possible pairs
+        for reactant_pair, product_pair in itertools.product(reactant_pairs, product_pairs):
+            try:
+                # see if we obtain proper labeling
+                labeled_reactants, labeled_products = self.getLabeledReactantsAndProducts(reactant_pair, product_pair)
+                if labeled_reactants is not None:
+                    break
+            except ActionError:
+                # must have gotten the wrong pair
+                pass
+        if labeled_reactants is None or labeled_products is None:
+            raise ActionError("Could not find labeled reactants for reaction {} from family {}.".format(reaction,self.label))
+
+        # place the molecules in reaction object
+        reaction.products = labeled_products
+        reaction.reactants = labeled_reactants
+
+        if output_with_species:
```

Comment:
Is there a use case where output with molecule is preferred?

I think a resonance argument to toggle resonance structure generation might be more useful.

Reply:
In my isotope work, I use the output molecule objects for some molecular graph operations. 

I also like the idea of only outputting species objects. One potential issue I realized with this method would be the loss of any species-level data (like thermo, symmetry number, etc) when using this method. 

I propose that I make the following changes to the method:

1. only output species objects with the option to create resonance structures
2. store the updated & labeled graphs in the current species objects to prevent loss of any species-level attributes.

How does this sound?

Reply:
I think that sounds good. Will using the original species objects be straightforward to implement?